micropython/lora/lora/modem.py

```diff
@@ -0,0 +1,510 @@
+    # callback, record a timestamp, and wake up the hardware if
+    # needed. ppplication code is expected to call poll_transmit() or
+    # poll_receive() as applicable in order to confirm the modem state.
+    #
+    # This is a MP hard irq in some configurations, meaning no memory allocation is possible.
+    #
+    # 'pin' may also be None if this is a "soft" IRQ triggered after a receive
+    # timed out during a transmit (meaning no receive IRQ will fire, but the
+    # receiver should wake up and move on anyhow.)
+    def _radio_isr(self, pin):
+        self._last_irq = time.ticks_ms()
+        if self._irq_callback:
+            self._irq_callback(pin)
+        if _DEBUG:
+            # Note: this may cause a MemoryError and fail if _DEBUG is enabled in this base class
+            # but disabled in the subclass, meaning this is a hard irq handler
+            try:
+                print("_radio_isr pin={}".format(pin))
+            except MemoryError:
+                pass
+
+    def irq_triggered(self):
+        # Returns True if the ISR has executed since the last time a send or a receive
+        # started
+        return self._last_irq is not None
+
+    def set_irq_callback(self, callback):
+        # Set a function to be called from the radio ISR
+        #
+        # This is used by the AsyncModem implementation, but can be called in
+        # other circumstances to implement custom ISR logic.
+        #
+        # Note that callback may be called in hard ISR context, meaning no
+        # memory allocation is possible.
+        self._irq_callback = callback
+
+    def _get_last_irq(self):
+        # Return the _last_irq timestamp if set by an ISR, or the
+        # current time.time_ms() timestamp otherwise.
+        if self._last_irq is None:
+            return time.ticks_ms()
+        return self._last_irq
+
+    # Common parts of receive API
+
+    def start_receive(self, timeout_ms=None, continuous=False, rx_length=0xFF):
+        # Start receiving.
+        #
+        # Part of common low-level modem API, see README.md for usage.
+        if continuous and timeout_ms is not None:
+            raise ValueError()  # these two options are mutually exclusive
+
+        if timeout_ms is not None:
+            self._rx = (time.ticks_add(time.ticks_ms(), timeout_ms), continuous, rx_length)
+        else:
+            self._rx = (None, continuous, rx_length)
+
+    def poll_receive(self, rx_packet=None):
+        # Should be called while a receive is in progress:
+        #
+        # Part of common low-level modem API, see README.md for usage.
+        #
+        # This function may alter the state of the modem - it will clear
+        # RX interrupts, it may read out a packet from the FIFO, and it
+        # may resume receiving if the modem has gone to standby but receive
+        # should resume.
+
+        if not self._rx:
+            # Not actually receiving...
+            return (False, None)
```

Comment:
@dpgeorge @jimmo In the receive and transmit control flows I now have this pattern of returning 2-tuples and 3-tuples in a few places. In the previous version of this PR these types of results were constrained to "small" types but it got fiddly, not to mention quite confusing to read.

What's your feeling on these additional small heap allocations? I'm thinking I could rework it to allocate a set of list objects and keep them around for reuse, like what's done for the modem command buffers already, but I'm not certain it's worth the complexity vs RAM use tradeoff.

FWIW I think these are the only places that heap allocations can't be avoided during the normal transmit or receive control flow.

Reply:
Following some discussion offline, I've removed these tuples and the normal send/receive code paths are heap allocation free (provided rx_packet instance is passed into receive() and the new packet has the same length.)